src/conn/conn_api.c

```diff
@@ -2042,11 +2042,7 @@ __wt_debug_mode_config(WT_SESSION_IMPL *session, const char *cfg[])
         FLD_SET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN);
```

Comment:
This function `__wt_debug_mode_config` is accessed in two places: `__wt_conn_reconfig` which holds `conn->reconfig_lock`, and `wiredtiger_open` which doesn't hold the lock.
Two points to discuss:
- Should we look at taking the `reconfig_lock` during `wiredtiger_open`? We know for a fact we're the only thread at this point but it'd help with correctness checks if we always hold the lock when changing `debug_log_cnt`
- How can we know a lock is held when we enter a function? Some functions have a "Must be called with the XXX lock held" comment but is this enough?

Reply:
To quickly answer your second questions see @pmacko86's work on spinlock assertions: https://github.com/wiredtiger/wiredtiger/pull/9049/files.